### PR TITLE
Check compatibility

### PR DESCRIPTION
**Pull request by minsulee2**
In order to check and find a bug of data compatibility, some updates are made based on AIP-13.

Notes:
- Tidy up some checks
- Be the same as AIP-13 spec
- Pass compatibility check via local test

### CODE REVIEW

File: p2p/index.js [hunk 1]
```diff
@@ -342,9 +326,23 @@ class P2pClient {
           }
           break;
         case MessageTypes.CHAIN_SEGMENT_RESPONSE:
-          if (!this.checkDataProtoVerForChainSegmentResponse(parsedMessage.dataProtoVer)) {
+          const majorVersion = VersionUtil.toMajorVersion(parsedMessage.dataProtoVer);
```
Comment (liayoo) [inline p2p/index.js:329]: Is there a reason why you removed the checkDataProtoVerForChainSegmentResponse() and moved the code here? I feel like it was cleaner & more modularized before..

**Comment by liayoo** [comment]
It'd be really nice if we could see my version & the received/expected version in version-related error logs!

File: p2p/index.js [hunk 2]
```diff
@@ -342,9 +326,23 @@ class P2pClient {
...
+          const isLower = semver.lt(this.server.majorDataProtocolVersion, majorVersion);
```
Comment (platfowner) [inline p2p/index.js:330]: It seems this pattern is repeated in this file. What about extracting it as a separate function? e.g. in P2pServer:
```
compareWithDataProtoVersion(protoVersion) {
  const majorVersion = VersionUtil.toMajorVersion(protoVersion);
  if (semver.lt(majorVersion, this.majorDataProtocolVersion)) {
    return -1;
  }
  if (semver.gt(majorVersion, this.majorDataProtocolVersion)) {
    return 1;
  }
  return 0;
}
```

File: p2p/index.js (hunk 1 above)
Comment (platfowner) [reply p2p/index.js:329]: +1

**Comment by platfowner** [comment]
As discussed offline, let's test with updated policy (AIP-013) and see how it works. Thanks! @minsulee2 @liayoo

File: p2p/index.js (hunk 1 above)
Comment (minsulee2) [reply p2p/index.js]: Oh, right! All fixed like before.

File: p2p/index.js (hunk 2 above)
Comment (minsulee2) [reply p2p/index.js]: Done!

**Comment by minsulee2** [comment]
PTAL! I fixed all and the experiment was succeed!

File: p2p/util.js [hunk 3]
```diff
@@ -113,6 +116,27 @@ function checkTimestamp(timestamp) {
...
+      logger.error(`The given ${msgType} message is stale.`);
+    }
+    return 1;
+  }
+  const isLower = semver.lt(nodeMajorVersion, comingMajorVersion);
+  if (isLower) {
+    if (FeatureFlags.enableRichP2pCommunicationLogging) {
+      logger.error('I may be running of the old DATA_PROTOCOL_VERSION of ain-blockchain node. ' +
+          'Please check the new release via visiting the URL below:');
+      logger.error('https://github.com/ainblockchain/ain-blockchain');
```
Comment (platfowner) [inline p2p/util.js:133]: Could you combine the two lines above? maybe with '... URL below:\n...'?

File: p2p/util.js [hunk 4]
```diff
@@ -113,6 +116,27 @@ function checkTimestamp(timestamp) {
   }
 }
 
+function checkDataProtoVer(nodeMajorVersion, comningVersion, msgType) {
```
Comment (platfowner) [inline p2p/util.js:119]: type? 'comningVersion'

Comment (platfowner) [reply p2p/util.js:119]: minor: what about 'messageVersion'?

Comment (platfowner) [inline p2p/util.js:119]: minor: would it be more natural to add this function as a member of P2pServer as it always use                 this.majorDataProtocolVersion?

File: p2p/util.js [hunk 5]
```diff
@@ -113,6 +116,27 @@ function checkTimestamp(timestamp) {
...
+  const comingMajorVersion = VersionUtil.toMajorVersion(comningVersion);
+  const isGreater = semver.gt(nodeMajorVersion, comingMajorVersion);
```
Comment (platfowner) [inline p2p/util.js:121]: minor: would it be more natural to use `semver.lt(comingMajorVersion, nodeMajorVersion)` as this function is more about `comningVersion` than `nodeMajorVersion`?

File: p2p/util.js [hunk 6]
```diff
@@ -113,6 +116,27 @@ function checkTimestamp(timestamp) {
...
+  if (isGreater) {
+    if (FeatureFlags.enableRichP2pCommunicationLogging) {
+      logger.error(`The given ${msgType} message is stale.`);
+    }
+    return 1;
+  }
+  const isLower = semver.lt(nodeMajorVersion, comingMajorVersion);
```
Comment (platfowner) [inline p2p/util.js:128]: ditto

File: p2p/util.js [hunk 7]
```diff
@@ -113,6 +116,27 @@ function checkTimestamp(timestamp) {
...
 }
 
+function checkDataProtoVer(nodeMajorVersion, comningVersion, msgType) {
+  const comingMajorVersion = VersionUtil.toMajorVersion(comningVersion);
+  const isGreater = semver.gt(nodeMajorVersion, comingMajorVersion);
+  if (isGreater) {
+    if (FeatureFlags.enableRichP2pCommunicationLogging) {
+      logger.error(`The given ${msgType} message is stale.`);
+    }
+    return 1;
```
Comment (platfowner) [inline p2p/util.js:126]: minor: what about returning -1 in this case as this function is more about `comningVersion`?

File: p2p/server.js [hunk 8]
```diff
@@ -438,19 +395,32 @@ class P2pServer {
...
             break;
           case MessageTypes.TRANSACTION:
+            const dataVersionCheckForTransaction = checkDataProtoVer(
+                this.majorDataProtocolVersion, dataProtoVer, 'TRANSACTION');
+            if (dataVersionCheckForTransaction < 0) {
+              logger.error('CANNOT deal with higher data protocol version. Discard the ' +
+                  'TRANSACTION message.');
+              return;
+            }
+            else if (dataVersionCheckForTransaction > 0) {
```
Comment (platfowner) [inline p2p/server.js:420]: minor: could this be merged with the above line?

File: p2p/index.js (hunk 2 above)
Comment (platfowner) [reply p2p/index.js]: Thanks!

**Comment by platfowner** [comment]
> It'd be really nice if we could see my version & the received/expected version in version-related error logs!

I think it would be very helpful for the node runners. What do you think? @minsulee2

**Comment by platfowner** [comment]
minor: I see some case that one message is divided with multiple calls of `logger.error()` e.g. in server.js:
```
            logger.error(`Wrong message type(${parsedMessage.type}) has been specified.`);
            logger.error('Ignore the message.')
```
could you merge them into one log message? I would be helpful to debug with e.g. stack driver.

File: p2p/index.js [hunk 9]
```diff
@@ -344,15 +317,22 @@ class P2pClient {
           }
           break;
         case MessageTypes.CHAIN_SEGMENT_RESPONSE:
+          const dataVersionCheckForChainSegment = checkDataProtoVer(
```
Comment (liayoo) [inline p2p/index.js:320]: How about we perform this check after the node state check? If the node is starting, it's going to ignore the message anyway, so..

File: p2p/util.js [hunk 10]
```diff
@@ -113,6 +116,27 @@ function checkTimestamp(timestamp) {
   }
 }
 
+function checkDataProtoVer(nodeMajorVersion, comningVersion, msgType) {
+  const comingMajorVersion = VersionUtil.toMajorVersion(comningVersion);
+  const isGreater = semver.gt(nodeMajorVersion, comingMajorVersion);
+  if (isGreater) {
+    if (FeatureFlags.enableRichP2pCommunicationLogging) {
+      logger.error(`The given ${msgType} message is stale.`);
```
Comment (liayoo) [inline p2p/util.js:124]: minor) I think "stale" makes the reader think that the message's timestamp is old. How about something like `The given ${msgType} message has unsupported DATA_PROTOCOL_VERSION: theirs (${comningVersion}) < ours (${nodeMajorVersion})` ?

File: p2p/util.js (hunk 3 above)
Comment (minsulee2) [reply p2p/util.js]: Fixed!

File: p2p/util.js (hunk 4 above)
Comment (minsulee2) [reply p2p/util.js]: I got 'messageVersion' Thanks!

Comment (minsulee2) [reply p2p/util.js]: Yeah, it looks way better. As changed the code a lot, the code was a bit bizarre. Fixed!

File: p2p/util.js (hunk 5 above)
Comment (minsulee2) [reply p2p/util.js]: Thought the same way and fixed! It is way better and intuitive.

File: p2p/util.js (hunk 6 above)
Comment (minsulee2) [reply p2p/util.js]: Done!

File: p2p/util.js (hunk 7 above)
Comment (minsulee2) [reply p2p/util.js]: Done!

**Comment by minsulee2** [comment]
> > It'd be really nice if we could see my version & the received/expected version in version-related error logs!
> 
> I think it would be very helpful for the node runners. What do you think? @minsulee2



> > It'd be really nice if we could see my version & the received/expected version in version-related error logs!
> 
> I think it would be very helpful for the node runners. What do you think? @minsulee2

Oh! I think so too. I will update it as it is

**Comment by minsulee2** [comment]
> minor: I see some case that one message is divided with multiple calls of `logger.error()` e.g. in server.js:
> 
> ```
>             logger.error(`Wrong message type(${parsedMessage.type}) has been specified.`);
>             logger.error('Ignore the message.')
> ```
> 
> could you merge them into one log message? I would be helpful to debug with e.g. stack driver.

Done!!

File: p2p/index.js (hunk 9 above)
Comment (minsulee2) [reply p2p/index.js]: Done!

File: p2p/util.js (hunk 10 above)
Comment (minsulee2) [reply p2p/util.js]: Applied! thanks! This seems intuitive much!

File: p2p/server.js (hunk 8 above)
Comment (minsulee2) [reply p2p/server.js:438]: Fixed!

**Comment by minsulee2** [comment]
PTAL. All addressed. I will move on to an experiment again with updated codes.

File: p2p/server.js [hunk 11]
```diff
@@ -390,8 +366,11 @@ class P2pServer {
 
         switch (_.get(parsedMessage, 'type')) {
           case MessageTypes.ADDRESS_REQUEST:
-            // TODO(minsulee2): Add compatibility check here after data version up.
-            // this.checkDataProtoVerForAddressRequest(dataProtoVer);
+            const dataVersionCheckForAddress = this.checkDataProtoVer(dataProtoVer, 'ADDRESS_REQUEST');
```
Comment (liayoo) [inline p2p/server.js:369]: MessageTypes.ADDRESS_REQUEST

File: p2p/server.js [hunk 12]
```diff
@@ -438,19 +417,29 @@ class P2pServer {
             }
             break;
           case MessageTypes.CONSENSUS:
+            const dataVersionCheckForConsensus = this.checkDataProtoVer(dataProtoVer, 'CONSENSUS');
```
Comment (liayoo) [inline p2p/server.js:420]: MessageTypes.CONSENSUS

File: p2p/server.js [hunk 13]
```diff
@@ -438,19 +417,29 @@ class P2pServer {
...
-              return;
-            }
             if (this.node.state === BlockchainNodeStates.SERVING) {
               this.consensus.handleConsensusMessage(consensusMessage);
             } else {
               logger.info(`\n [${LOG_HEADER}] Needs syncing...\n`);
             }
             break;
           case MessageTypes.TRANSACTION:
+            const dataVersionCheckForTransaction = checkDataProtoVer(dataProtoVer, 'TRANSACTION');
```
Comment (liayoo) [inline p2p/server.js:434]: MessageTypes.TRANSACTION

File: p2p/index.js [hunk 14]
```diff
@@ -309,8 +277,11 @@ class P2pClient {
 
       switch (parsedMessage.type) {
         case MessageTypes.ADDRESS_RESPONSE:
-          // TODO(minsulee2): Add compatibility check here after data version up.
-          // this.checkDataProtoVerForAddressResponse(dataProtoVer);
+          const dataVersionCheckForAddress = this.server.checkDataProtoVer(dataProtoVer, 'ADDRESS_RESPONSE');
```
Comment (liayoo) [inline p2p/index.js:280]: How about using MessageTypes.ADDRESS_RESPONSE instead of 'ADDRESS_RESPONSE'?

File: p2p/index.js [hunk 15]
```diff
@@ -350,8 +321,14 @@ class P2pClient {
                 `Node state: ${this.server.node.state}.`);
             return;
           }
-          if (!this.checkDataProtoVerForChainSegmentResponse(parsedMessage.dataProtoVer)) {
+          const dataVersionCheckForChainSegment = this.server.checkDataProtoVer(dataProtoVer, 'CHAIN_SEGMENT_RESPONSE');
```
Comment (liayoo) [inline p2p/index.js:324]: MessageTypes.CHAIN_SEGMENT_RESPONSE

**Review by liayoo — APPROVED** [review]
LGTM with some comments. Thanks! 🙌

File: p2p/index.js [hunk 16]
```diff
@@ -350,8 +322,15 @@ class P2pClient {
                 `Node state: ${this.server.node.state}.`);
             return;
           }
-          if (!this.checkDataProtoVerForChainSegmentResponse(parsedMessage.dataProtoVer)) {
+          const dataVersionCheckForChainSegment =
+              this.server.checkDataProtoVer(dataProtoVer, MessageTypes.CHAIN_SEGMENT_RESPONSE);
+          if (dataVersionCheckForChainSegment > 0) {
+            logger.error('CANNOT deal with higher data protocol version. Discard the ' +
```
Comment (platfowner) [inline p2p/index.js:328]: minor: versions missing from the error message?

File: p2p/server.js [hunk 17]
```diff
@@ -438,19 +418,31 @@ class P2pServer {
...
               this.consensus.handleConsensusMessage(consensusMessage);
             } else {
               logger.info(`\n [${LOG_HEADER}] Needs syncing...\n`);
             }
             break;
           case MessageTypes.TRANSACTION:
+            const dataVersionCheckForTransaction =
+                this.checkDataProtoVer(dataProtoVer, MessageTypes.TRANSACTION);
+            if (dataVersionCheckForTransaction > 0) {
+              logger.error('CANNOT deal with higher data protocol version. Discard the ' +
```
Comment (platfowner) [inline p2p/server.js:439]: minor: versions missing from the error message?

File: p2p/server.js [hunk 18]
```diff
@@ -438,19 +418,31 @@ class P2pServer {
             }
             break;
           case MessageTypes.CONSENSUS:
+            const dataVersionCheckForConsensus =
+                this.checkDataProtoVer(dataProtoVer, MessageTypes.CONSENSUS);
+            if (dataVersionCheckForConsensus !== 0) {
+              return;
```
Comment (platfowner) [inline p2p/server.js:424]: what about leaving some error messages like other cases (note that it no longer allows further handling logic in handleConsensusMessage() like before)?

File: p2p/server.js [hunk 19]
```diff
@@ -529,8 +516,8 @@ class P2pServer {
             }
             break;
           default:
-            logger.error(`Wrong message type(${parsedMessage.type}) has been specified.`);
-            logger.error('Ignore the message.');
+            logger.error(`Wrong message type(${parsedMessage.type}) has been specified.` +
```
Comment (platfowner) [inline p2p/server.js:519]: minor: what about 'Wrong' -> 'Unknown' or something?

File: p2p/server.js (hunk 11 above)
Comment (minsulee2) [reply p2p/server.js]: Done!

File: p2p/server.js (hunk 12 above)
Comment (minsulee2) [reply p2p/server.js]: Done!

File: p2p/server.js (hunk 13 above)
Comment (minsulee2) [reply p2p/server.js]: Done!

File: p2p/index.js (hunk 14 above)
Comment (minsulee2) [reply p2p/index.js]: Done!

File: p2p/index.js (hunk 15 above)
Comment (minsulee2) [reply p2p/index.js]: Done!

File: p2p/index.js (hunk 16 above)
Comment (minsulee2) [reply p2p/index.js]: Done!

File: p2p/server.js (hunk 17 above)
Comment (minsulee2) [reply p2p/server.js]: Done!

File: p2p/server.js (hunk 18 above)
Comment (minsulee2) [reply p2p/server.js:426]: Done!

File: p2p/server.js (hunk 19 above)
Comment (minsulee2) [reply p2p/server.js]: Done!

**Comment by minsulee2** [comment]
PTAL. Addressed the minor comments!

**Review by platfowner — APPROVED** [review]
LGTM
Thanks for the nice refactoring!

**Comment by minsulee2** [comment]
Thanks for the review! All experiments passed. Merging now.

**Pull request merged**